### PR TITLE
ISPN-6218 Port ClusterRoleMapper to InternalCacheRegistry

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-6218

### CODE REVIEW

**Comment by anistor** [comment]
Looking ..

File: core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java [hunk 1]
```diff
@@ -1,16 +1,21 @@
...
 import org.infinispan.security.PrincipalRoleMapperContext;
 
+import java.security.Principal;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.Set;
+
 /**
  * ClusterPrincipalMapper.
```
Comment (anistor) [inline core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java]: ClusterRoleMapper?

File: core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java [hunk 2]
```diff
@@ -19,12 +24,22 @@
  */
 public class ClusterRoleMapper implements PrincipalRoleMapper {
    private EmbeddedCacheManager cacheManager;
-   private ClusterRegistry<Class<?>, String, Set<String>> clusterRegistry;
+   private static final String CLUSTER_ROLE_MAPPER_CACHE = "___clusterrolemapper";
```
Comment (anistor) [inline core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java]: ```___cluster_role_mapper``` maybe?

File: core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java [hunk 3]
```diff
@@ -34,29 +49,29 @@
...
+      InternalCacheRegistry internalCacheRegistry = cacheManager.getGlobalComponentRegistry().getComponent(InternalCacheRegistry.class);
+      internalCacheRegistry.registerInternalCache(CLUSTER_ROLE_MAPPER_CACHE, cfg.build(), EnumSet.of(InternalCacheRegistry.Flag.PERSISTENT));
    }
 
    public void grant(String roleName, String principalName) {
-      Set<String> roleSet = clusterRegistry.get(ClusterRoleMapper.class, principalName);
-      if (roleSet == null) {
-         roleSet = new HashSet<String>();
-      }
+      Set<String> roleSet = getClusterRoleMap().computeIfAbsent(principalName, n -> { return new HashSet<>(); });
```
Comment (anistor) [inline core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java]: replace ```{ return new HashSet<>(); }``` with ```new HashSet<>()``` ?

File: core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java [hunk 4]
```diff
@@ -34,29 +49,29 @@
...
-      clusterRegistry.put(ClusterRoleMapper.class, principalName, roleSet);
+      clusterRoleMap.put(principalName, roleSet);
    }
 
    public void deny(String roleName, String principalName) {
-      Set<String> roleSet = clusterRegistry.get(ClusterRoleMapper.class, principalName);
-      if (roleSet == null) {
-         roleSet = new HashSet<String>();
-      }
+      Set<String> roleSet = getClusterRoleMap().computeIfAbsent(principalName, n -> { return new HashSet<>(); });
```
Comment (anistor) [inline core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java]: same here

File: core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java (hunk 3 above)
Comment (wburns) [inline core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java]: Should be able to replace the entire thing with ``HashSet::new``

Comment (anistor) [inline core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java]: Indeed.

**Comment by tristantarrant** [comment]
I have implemented all suggestions

File: core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java (hunk 3 above)
Comment (wburns) [inline core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java]: Whoops forgot computeIfAbsent passes the key as an argument, disregard my comment :-1:

File: core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java [hunk 5]
```diff
@@ -34,29 +49,29 @@
...
-      clusterRegistry.put(ClusterRoleMapper.class, principalName, roleSet);
+      clusterRoleMap.put(principalName, roleSet);
    }
 
    public Set<String> list(String principalName) {
-      Set<String> roleSet = clusterRegistry.get(ClusterRoleMapper.class, principalName);
+      Set<String> roleSet = getClusterRoleMap().get(principalName);
       if (roleSet != null) {
          return Collections.unmodifiableSet(roleSet);
       } else {
```
Comment (anistor) [inline core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java]: Do we still need ```InfinispanCollections.emptySet()``` ? I believe all ```InfinispanCollections.empty*()``` should be deprecated now.

Comment (tristantarrant) [inline core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java]: Go ahead with a PR for that :)

File: core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java [hunk 6]
```diff
@@ -34,29 +49,29 @@
    @Override
    public void setContext(PrincipalRoleMapperContext context) {
       this.cacheManager = context.getCacheManager();
-      clusterRegistry = cacheManager.getGlobalComponentRegistry().getComponent(ClusterRegistry.class);
+      GlobalConfiguration globalConfiguration = cacheManager.getGlobalComponentRegistry().getGlobalConfiguration();
+      CacheMode cacheMode = globalConfiguration.isClustered() ? CacheMode.REPL_SYNC : CacheMode.LOCAL;
+      ConfigurationBuilder cfg = new ConfigurationBuilder();
```
Comment (anistor) [inline core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java]: The ClusterRegistry also disabled authorization for internal caches, which no longer happens here. Was that an unwanted side effect?

Comment (tristantarrant) [inline core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java]: I've updated the PR to do the same.

File: core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java (hunk 5 above)
Comment (anistor) [inline core/src/main/java/org/infinispan/security/impl/ClusterRoleMapper.java]: I've added a jira for this https://issues.jboss.org/browse/ISPN-6226

**Comment by anistor** [comment]
Integrated. Thanks @tristantarrant !

**Pull request merged**